Full recovery mode

ztlpn opened the pull request:
Add "full recovery mode":
* don't load user partitions
* disable balancers; pandaproxy and schema registry listeners
* return errors from produce/consume-related kafka API handlers

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes

### Features

* Add *recovery mode* - an option (for the purposes of disaster recovery) to start redpanda in "metadata-only" mode, skipping loading user partitions and allowing only metadata operations. Enabled by the `recovery_mode_enabled` node config property.

vbotbuildovich commented:
new failures detected in https://buildkite.com/redpanda/redpanda/builds/39141#018b3fb7-001b-4617-b94f-da3a4fe4e894: "rptest.tests.tiered_storage_model_test.TieredStorageTest.test_tiered_storage.cloud_storage_type=CloudStorageType.S3.test_case=.TS_Read==True.TS_TxRangeMaterialized==True.SpilloverManifestUploaded==True"

ztlpn commented:
The error doesn't look related to my changes, opened https://github.com/redpanda-data/redpanda/issues/14266

bharathv commented on src/v/kafka/server/server.cc:
```diff
@@ -704,6 +709,12 @@ add_offsets_to_txn_handler::handle(request_context ctx, ss::smp_service_group) {
         request.decode(ctx.reader(), ctx.header().version);
         log_request(ctx.header(), request);
 
+        if (ctx.recovery_mode_enabled()) {
```
we may also want to add the check in `init_producer_id_handler::handle`, client uses it to get a producer_id to begin a idempotency  / transaction session.

ztlpn replied:
It's there! https://github.com/redpanda-data/redpanda/pull/14236/files#diff-d66a9af30ebd2f5bc30d59587cb0f9b1d0d8ea3121a9d5411a08d69191319af4R1138

bharathv reviewed (commented):
lgtm, all nits/clarification.

ztlpn commented on src/v/cluster/controller_backend.cc:
```diff
@@ -769,7 +775,14 @@ void controller_backend::housekeeping() {
+ss::future<>
+controller_backend::reconcile_ntp(const model::ntp& ntp, deltas_t& deltas) {
+    if (should_skip(ntp)) {
+        vlog(clusterlog.info, "[{}] skipping reconcilation", ntp);
```
good point

ztlpn commented on src/v/kafka/server/request_context.h:
```diff
@@ -294,6 +298,7 @@ class request_context {
     request_header _header;
     protocol::decoder _reader;
     ss::lowres_clock::duration _throttle_delay;
+    bool _recovery_mode_enabled = false;
```
Well, several points:
* It is generally better to make request handling more self-contained and less reliant on thread-local singletons (might help with testing for example)
* It is used by more than 1/3 of handlers, so looks like this is a relevant bit of information that should be readily accessible by request handlers. It also serves as a bit of a reminder for handler authors to at least think about checking it.
* It is shorter :)

mmaslankaprv replied:
maybe we can make the `_recovery_mode_enabled` part of the `kafka::server` api, this way a `request_ctx` would proxy a call to the server.

ztlpn replied:
right, looks like there are other flags exposed this way

mmaslankaprv commented on src/v/model/namespace.h:
```diff
@@ -75,4 +75,14 @@ inline const model::topic_partition schema_registry_internal_tp{
 inline const model::ntp wasm_binaries_internal_ntp(
   model::redpanda_ns, model::topic("wasm_binaries"), model::partition_id(0));
 
+inline bool is_user_topic(const topic_namespace_view& tp_ns) {
```
the const ref is not required, it is already a view

ztlpn replied:
fixed

mmaslankaprv commented on src/v/kafka/server/handlers/produce.cc:
```diff
@@ -632,6 +632,12 @@ produce_handler::handle(request_context ctx, ss::smp_service_group ssg) {
     request.decode(ctx.reader(), ctx.header().version);
     log_request(ctx.header(), request);
+
+    if (ctx.recovery_mode_enabled()) {
```
nit: since this is a performance critical path we may use `unlikely`

ztlpn replied:
fixed

mmaslankaprv commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -879,6 +879,10 @@ fetch_handler::handle(request_context rctx, ss::smp_service_group ssg) {
                   octx.response.data.error_code = octx.session_ctx.error();
                   return std::move(octx).send_response();
               }
+              if (octx.rctx.recovery_mode_enabled()) {
```
`unlikely` ?

mmaslankaprv commented on src/v/kafka/server/server.cc:
```diff
@@ -366,6 +366,11 @@ ss::future<response_ptr> heartbeat_handler::handle(
     request.decode(ctx.reader(), ctx.header().version);
     log_request(ctx.header(), request);
 
+    if (ctx.recovery_mode_enabled()) {
```
this and offset commit are performance critical code paths please wrap with `unlikely`

ztlpn commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/303c13ffe240b0eb57a0ab8d310afc8589c8e53f..354f06bc82ed23c26c166373ea9b62838ca3261d): addressed review comments and added group describe test checks in recovery mode.

ztlpn commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/303c13ffe240b0eb57a0ab8d310afc8589c8e53f..9a409eb0ad52e865f9fcdd222d9af932dc1a8ddd): addressed review comments and added group describe test checks in recovery mode.

emaxerrno commented:
@ztlpn this is coooool! should this come w/ a list of admin api endpoints that are avail ... perhaps admin api endpoints to trigger GC of segments or smth like that.

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/39441#018b4dee-f6e7-4c66-9f2e-1a6cc2a0ffcc

mmaslankaprv commented:
/ci-repeat 1

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/39570#018b5be3-0fc3-405f-9b4a-efaf08755d8f

ztlpn commented:
@emaxerrno Metadata operations are still available so most of the existing admin API should work without problems. Having additional endpoints for fixing problematic partitions makes sense but is a bit out of scope for this project (as a start, it would be great to at least have an ability to delete them and recovery mode allows this).
